services/indexing/search.go

```diff
@@ -171,8 +191,11 @@ func (self *Indexer) getAllClients(
 	}
 
 	terms := []string{allClientsQuery}
-	return self.searchWithTerms(ctx, config_obj,
+	clients, _, err := self.searchWithTerms(ctx, config_obj,
```

Comment:
This actually needs to reflect the total number of hits not the total number of clients. The GUI uses this to create paging for the hits so say if only 2 clients match the search but there are 10k clients in the index the table will be way off.

I think the total number of matches is actually already returned as part of the search query anyway we just need to also extract it from the JSON response.

Reply:
I don't think the actual total of clients matched is actually returned, it seems that the only performant way to determine a total with OpenSearch is with the [Count API](https://opensearch.org/docs/latest/api-reference/count/)

Otherwise it looks like you face pretty extreme performance loss when dealing with large numbers of documents (for example if we have orgs with ~100,000 clients etc)

Reply:
Not an elastic expert but the link you posted actually has an example using the search API:

```
The following call to the search API produces equivalent results:

GET opensearch_dashboards_sample_data_logs/_search
{
  "query": {
    "term": {
      "response": "200"
    }
  },
  "size": 0,
  "track_total_hits": true
}
```

I am not sure if this is in addition to the search results (e.g. if size is not 0 but track total hits is still set).

If we need to reissue the same query with the count API  we should do this in searchWithTerms and surface the total count back up since we dont want the total number of clients, only the total number of matching clients.

Reply:
I believe `track_total_hits` ends up being more costly than using the Count API when dealing with large volumes of docs, the OpenSearch docs don't seem to go into any detail on it but there is some discussion about it various GitHub issues and forums.

Count apparently should demonstrate value if we match against incredibly large numbers of documents since it doesn't do any ranking optimisations or fetch the `_source` for those documents (though it requires an additional HTTP req in our use case)

Reply:
This will return the total number of clients rather than the total number of OpenSearch documents. By only counting documents where the hostname is present we get the true total rather than then having to count all and divide that number by 4 or 5 to guess a true number. Count is also the better API for this because Search will hold the docs in memory during the query where as count will just increment an internal counter, and since we don't want to retrieve them all this is the better solution. The search API also tops out at 10,000 results and 69,000 aggregations, so we would have to call the API multiple times to get the same count we can get with one call to the _count api.

Reply:
There are clearly some subtleties around documents and clients, can we add a test to be sure this works?

Reply:
> There are clearly some subtleties around documents and clients, can we add a test to be sure this works?

Yeah I've tested it locally with Docker containers but I could add a unit test to [ingestor_test.go](https://github.com/Velocidex/cloudvelo/blob/master/ingestion/ingestor_test.go) if you think thats sufficient for testing the new request to the [Count API](https://github.com/Velocidex/cloudvelo/pull/149/files#diff-442912520850883d6918973859f6e0e0a9891c285ed32f77c4bb37731a9821c1R883)

Reply:
Yeah i see that we dont have a test for this part of the code. It is not related to the ingestor so probably not ideal to add the test there. We should replicate the test suite in this service as well.

 Lets submit it now and we can think about adding more tests later.